api.go

```diff
@@ -122,6 +122,42 @@ func NewApiWithClient(a ma.Multiaddr, c *gohttp.Client) (*HttpApi, error) {
 	return api, nil
 }
 
+// NewDirectAPIClient is used to instantiate a HttpApi client
+// that connects to an endpoint which leverages additional http paths.
+//
+// If you need to connect to a IPFS HTTP API located at https://foo.bar/baz/api/v0
+// you should use NewDirectAPIClient.
+func NewDirectAPIClient(url string) (*HttpApi, error) {
```

Comment:
Maybe `NewURLApi` to be consistent? (we need to make multiaddrs support this but that's an issue for another day).

Reply:
Also, can we avoid duplicating this code? E.g., pull out a helper function and make `NewApiWithClient` call that (first converting the multiaddr to a url)?

Reply:
Actually, we could probably just make this `NewURLApiWithClient` and then drop the `WithAuthorization` function.

Reply:
Sounds good I'll implement that today

Reply:
Re-reading this before I make the changes one thing I want to clarify, based off your [last](https://github.com/ipfs/go-ipfs-http-client/pull/15#discussion_r276533662) message remove `WithAuthorization` to allow for handling of headers as per the other conversation down below :arrow_down:, and change the name of the function to `NewURLApiWithClient`?

Reply:
That's how I read it.

Reply:
Perfect, I'll go ahead and make that change :ok_hand:

Reply:
Sorry for the delay, part of the issue I'm encountering making this change while trying to avoid touching the entire code base, is how to pass in the user given headers without:

a) overwritting the headers generated during request building
b) modifying every single instance of the request building function signature.

I don't believe `b` is possible, however this is my idea as to how to accomplish `a`

Modfying the `Exec` function to accept additional arguments of type `http.Header`

```Golang
// Exec sends the request a request and decodes the response.
func (r *RequestBuilder) Exec(ctx context.Context, additionalHeaders http.Header, res interface{}) error {
	httpRes, err := r.Send(ctx, additionalHeaders)
	if err != nil {
		return err
	}

	if res == nil {
		lateErr := httpRes.Close()
		if httpRes.Error != nil {
			return httpRes.Error
		}
		return lateErr
	}

	return httpRes.decode(res)
}
```

Then within `Send` checking to see if any additional headers were provided, and if so adding them to the existing headers

```Golang
// Send sends the request and return the response.
func (r *RequestBuilder) Send(ctx context.Context, additionalHeaders http.Header) (*Response, error) {
	r.shell.applyGlobal(r)

	req := NewRequest(ctx, r.shell.url, r.command, r.args...)
	req.Opts = r.opts
	req.Headers = r.headers
	req.Body = r.body
	if additionalHeaders != nil {
		for k := range additionalHeaders {
			req.Headers = additionalHeaders.Get(k)
		}
	}
	return req.Send(&r.shell.httpcli)
}
```

Reply:
You should be able to pass these from API to RequestBuilder when we construct the builder - https://github.com/ipfs/go-ipfs-http-client/blob/master/api.go#L142, without changing anything that's used everywhere

Reply:
Neat that is much better than my original idea :joy: 
```Golang
func (api *HttpApi) Request(command string, args ...string) *RequestBuilder {
	var headers map[string]string
	if api.Headers != nil {
		for k := range api.Headers {
			headers[k] = api.Headers.Get(k)
		}
	}
	return &RequestBuilder{
		command: command,
		args:    args,
		shell:   api,
		headers: headers,
	}
}
```

Seems to do the trick?

Reply:
Yep

Reply:
Okay I've pushed a fix that implements these changes, however tests are being wonky. I'll continue the test thread down below :arrow_down:  in the comment of your indicating that the daemon is no longer spun up in CI